[WiP]: Initial DDlog scaffolding in controller

amytai opened the pull request:
Opening a permanent PR to keep track of our ovn-controller changes.

This initial commit creates a dummy output relation and populates it programmatically.
Still todo:
- Create correct output relations
- Create adapters that populate DDlog whenever diffs are made to OVN_Southbound database (perhaps port from northd)

amytai committed:
Initial DDlog scaffolding in controller

- Create dummy output relation in ovn_controller.dl
- Populate this output relation programmatically inside ovn-controller main

amytai commented:
For now, the way to test whether things are working correctly is to run `make check TESTSUITEFLAGS='-j8 -d 924'` in the top-level directory. Then, inspect `tests/testsuite.dir/924/hv1/ovn-controller.log` and grep for `DDlog record`.

mihaibudiu commented on controller/ovn-controller.c:
```diff
@@ -2720,6 +2734,62 @@ main(int argc, char *argv[])
+
+    input_table = ddlog_get_table_id(prog, input_relation);
+    if (input_table == -1)
+        VLOG_INFO(" ddlog_get_table_id returned -1 for %s, %d\n", input_relation, input_table);
```
the main problem is that on all these errors you seemingly continue through the normal path as well.

blp replied:
I think this is just to get stuff building. We definitely would need better error handling for production but it's a start!

mihaibudiu commented on controller/ovn-controller.c:
```diff
@@ -2720,6 +2734,62 @@ main(int argc, char *argv[])
 
+    /*** Begin DDlog test code ***/
+    // DDlog vars
+    ddlog_prog prog;
```
I am not sure what this "main" is for. I wonder how control reaches this point.

blp replied:
This is the main code of ovn-controller. This new code will run as part of the daemon startup.

blp commented on controller/ovn-controller.c:
```diff
@@ -2562,6 +2564,18 @@ check_northd_version(struct ovsdb_idl *ovs_idl, struct ovsdb_idl *ovnsb_idl,
     return true;
 }
 
+static bool print_records_callback(uintptr_t arg, const ddlog_record *rec, ssize_t weight) {
```
The OVS/OVN style calls for the function name to start on its own line, and the `{` to be on its own line too, like this:

```
static bool
print_records_callback(uintptr_t arg, const ddlog_record *rec, ssize_t weight)
{
```

blp commented on controller/ovn-controller.c:
```diff
@@ -2562,6 +2564,18 @@ check_northd_version(struct ovsdb_idl *ovs_idl, struct ovsdb_idl *ovnsb_idl,
 }
 
+static bool print_records_callback(uintptr_t arg, const ddlog_record *rec, ssize_t weight) {
+    VLOG_INFO("Printing DDlog records\n");
```
Usually we don't write `\n` at the end of a log message, because the log system does that for us. (It doesn't hurt anything though.)

blp commented on controller/ovn-controller.c:
```diff
@@ -2720,6 +2734,62 @@ main(int argc, char *argv[])
+    }
+
+    input_table = ddlog_get_table_id(prog, input_relation);
+    if (input_table == -1)
```
I'm sure it doesn't matter for this test code, but OVS/OVN style calls for always putting `{}` around the statements gated by an `if`.

blp commented on controller/ovn-controller.c:
```diff
@@ -2562,6 +2564,18 @@ check_northd_version(struct ovsdb_idl *ovs_idl, struct ovsdb_idl *ovnsb_idl,
+    VLOG_INFO("Printing DDlog records\n");
+    char *record_as_string = ddlog_dump_record(rec);
+    if (record_as_string == NULL) {
+        VLOG_INFO(stderr, "failed to dump record\n");
```
The `stderr` here shouldn't be. My compiler complains, I'm not sure why yours doesn't?

```
../controller/ovn-controller.c:2571:9: error: incorrect type in argument 3 (different base types)
../controller/ovn-controller.c:2571:9:    expected char const *format
../controller/ovn-controller.c:2571:9:    got struct _IO_FILE [usertype] *extern [addressable] [toplevel] stderr
```

blp reviewed (commented):
I'm getting a lot of compiler warnings and errors here. All minor but should be fixed. You can use `OVS_UNUSED` to silence warnings about unused parameters.

```
../controller/ovn-controller.c: In function ‘print_records_callback’:
../controller/ovn-controller.c:2567:46: error: unused parameter ‘arg’ [-Werror=unused-parameter]
../controller/ovn-controller.c:2567:84: error: unused parameter ‘weight’ [-Werror=unused-parameter]
In file included from ../controller/ovn-controller.c:47:
../controller/ovn-controller.c: In function ‘main’:
../controller/ovn-controller.c:2752:19: error: format ‘%d’ expects argument of type ‘int’, but argument 5 has type ‘table_id’ {aka ‘long unsigned int’} [-Werror=format=]
/home/bpfaff/nicira/ovs/include/openvswitch/vlog.h:280:41: note: in definition of macro ‘VLOG’
../controller/ovn-controller.c:2752:9: note: in expansion of macro ‘VLOG_INFO’
../controller/ovn-controller.c:2752:61: note: format string is defined here
In file included from ../controller/ovn-controller.c:47:
../controller/ovn-controller.c:2756:19: error: format ‘%d’ expects argument of type ‘int’, but argument 5 has type ‘table_id’ {aka ‘long unsigned int’} [-Werror=format=]
/home/bpfaff/nicira/ovs/include/openvswitch/vlog.h:280:41: note: in definition of macro ‘VLOG’
../controller/ovn-controller.c:2756:9: note: in expansion of macro ‘VLOG_INFO’
../controller/ovn-controller.c:2756:60: note: format string is defined here
```

amytai committed:
Fix compiler warnings / formatting issues

amytai committed:
Add temporary hook in lflow for populating DDlog's Logical_Flow relation

This commit contains:
- new controller/ddlog.h header file to encapsulate ddlog boilerplate
- move insertion of DDlog input relation records to controller/lflow.c This might not be the permanent place: we need to identify the location to place the hook that writes OVSDB updates to the ovn-controller DDlog's Logical_Flow input relation.

Still TODO:
- consider_logical_flow logic should eventually be encapsulated in DDlog code
- find permanent location to insert OVSDB updates to ovn-controller DDlog instance

amytai committed:
Begin adding DDlog code that incrementally computes OF flows from OVN Logical Flows

Limitations:
- For now, only track address_set and port_group dependencies
- Don't have the logic for creating the actual OpenFlow flows
